src/audio_unit/macos_helpers.rs

```diff
@@ -161,12 +168,84 @@ pub fn get_audio_device_ids() -> Result<Vec<AudioDeviceID>, Error> {
     Ok(audio_devices)
 }
 
+pub fn get_audio_device_ids() -> Result<Vec<AudioDeviceID>, Error> {
+    return get_audio_device_ids_for_scope(Scope::Global);
+}
+
+/// does this device support input / ouptut?
+pub fn get_audio_device_supports_scope(devid: AudioDeviceID, scope: Scope) -> Result<bool, Error> {
+    let dev_scope: AudioObjectPropertyScope = match scope {
+        Scope::Input => kAudioObjectPropertyScopeInput,
+        Scope::Output => kAudioObjectPropertyScopeOutput,
+        _ => kAudioObjectPropertyScopeGlobal,
+    };
+    let property_address = AudioObjectPropertyAddress {
+        mSelector: kAudioDevicePropertyStreamConfiguration,
+        mScope: dev_scope,
+        mElement: kAudioObjectPropertyElementWildcard,
+    };
+
+    macro_rules! try_status_or_return {
+        ($status:expr) => {
+            if $status != kAudioHardwareNoError as i32 {
+                return Err(Error::Unknown($status));
+            }
+        };
+    }
+
+    let data_size = 0u32;
+    let status = unsafe {
+        AudioObjectGetPropertyDataSize(
+            devid,
+            &property_address as *const _,
+            0,
+            null(),
+            &data_size as *const _ as *mut _,
+        )
+    };
+    try_status_or_return!(status);
+
+    unsafe {
+        let buffers: *mut sys::AudioBufferList = libc::malloc(data_size as usize) as *mut sys::AudioBufferList;
```

Comment:
Other parts of this repo use vec u8 as a way to allocate some bytes (see render_callback.rs let mut data = vec![0u8; data_byte_size as usize]; but there is also a TODO there saying that it is leaking len and capacity fields of the vec.

A few answers here, https://stackoverflow.com/questions/45306575/how-can-you-allocate-a-raw-mutable-pointer-in-stable-rust the first answer suggests std::alloc.

On closer inspection, is it possible to remove libc and malloc/free by using a vec u8 `.as_mut_ptr()` and then you don't need to free because the vec will drop itself?